### PR TITLE
Improve scalability of StatsComponentSynchronizer and Boost

### PR DESCRIPTION
**Pull request by NancokPS2**
- Made it iterate trough an Array for easy addition/removal of stats in the future  
- Boost now supports arbitrary stats (no need for them to be pre-defined)

### CODE REVIEW

File: scripts/components/statssynchronizercomponent/StatsSynchronizerComponent.gd [hunk 1]
```diff
@@ -291,12 +312,15 @@ func add_experience(amount: int):
 
 
 func apply_boost(boost: Boost):
-	hp_max += boost.hp_max
-	hp += boost.hp
-	attack_power_min += boost.attack_power_min
-	attack_power_max += boost.attack_power_max
+	for statName in boost.statBoostDict:
+		assert(get(statName) != null, "This property does not exist as a stat.")
```
Comment (jonathaneeckhout) [inline scripts/components/statssynchronizercomponent/StatsSynchronizerComponent.gd:316]: We should not be using asserts on the server side. An assert will close the exe if it fails. This is a costly thing in an online service. It is better to handle the error cases by ignoring requests.

Comment (jonathaneeckhout) [inline scripts/components/statssynchronizercomponent/StatsSynchronizerComponent.gd:316]: I don't think this check should be in an assert but in an if statement with error handling to have it also present on released builds.

File: scripts/components/statssynchronizercomponent/StatsSynchronizerComponent.gd [hunk 2]
```diff
@@ -291,12 +312,15 @@ func add_experience(amount: int):
...
+		assert(
```
Comment (jonathaneeckhout) [inline scripts/components/statssynchronizercomponent/StatsSynchronizerComponent.gd:317]: same for this one

File: scripts/components/statssynchronizercomponent/StatsSynchronizerComponent.gd [hunk 3]
```diff
@@ -309,88 +333,46 @@ func apply_boost(boost: Boost):
...
 
-		if not "energy_max" in data:
-			GodotLogger.warn('Failed to load stats from data, missing "energy_max" key')
-			return false
+		data.merge(dictForMerge)
 
-		if not "attack_power_min" in data:
-			GodotLogger.warn('Failed to load stats from data, missing "attack_power_min" key')
-			return false
+		assert(data.size() == StatListCounter.size() + StatListPermanent.size())
```
Comment (jonathaneeckhout) [inline scripts/components/statssynchronizercomponent/StatsSynchronizerComponent.gd:349]: Same here. It's a valid check which should also be present on released builds.

File: scripts/components/statssynchronizercomponent/StatsSynchronizerComponent.gd [hunk 4]
```diff
@@ -291,12 +312,24 @@ func add_experience(amount: int):
...
+		if typeof(boost.get_stat_boost(statName)) != typeof(get(statName)):
+			(
+				GodotLogger
+				. warn(
+					(
+						"The value type of the boost ({0}) is different from the stat it is meant to alter ({1}). This could cause unexpected behaviour."
+						. format([typeof(boost.get_stat_boost(statName)), typeof(get(statName))])
+					)
+				)
+			)
```
Comment (NancokPS2) [inline scripts/components/statssynchronizercomponent/StatsSynchronizerComponent.gd:329]: gdformat goin crazy

**Comment by jonathaneeckhout** [comment]
Tested and works for me, thx for the contribution!

**Pull request merged**